compiler/src/iree/compiler/Dialect/Stream/Transforms/SpecializeEncodings.cpp

```diff
@@ -274,14 +275,143 @@ static RankedTensorType cloneWithEncoding(RankedTensorType type,
+      return;
+    }
+
+    // Bail out if the operation does not have an affinity attribute.
+    auto affinityAttr = affinityOp.getAffinityAttr();
+    if (!affinityAttr) {
+      return;
+    }
+    result.push_back(affinityOp);
+  });
+  return result;
+}
+
+namespace {
+
+// Adds the resolved layouts to all tensor types on stream tensor ops, if
+// encodings are present. Most of stream tensor ops implement
+// AffinityOpInterface, where a stream affinity indicates the kind of
+// enviroment the ops are expected run in. When an encoding is present in the
+// tensor type, the method resolves the layouts, strips outdated information,
+// and adds the resolved layouts to the encodings. The updated encodings should
+// have enough information for other lowering transformations.
+// TODO(hanchung): Add support for stream.tensor.load ops and
+// stream.tensor.store ops. They are not affinity ops, so additional analysis
+// will be needed in the work.
+class StreamTensorOpUpdater {
+public:
+  StreamTensorOpUpdater();
+  StreamTensorOpUpdater(ModuleOp moduleOp);
+  ~StreamTensorOpUpdater() {}
+
+  // Collects the stream tensor op candidates, and prepares all the needed
+  // information for the update. This must be called once before calling `run`.
+  // Note that all the ops are unmodified after the execution.
+  LogicalResult init();
+
+  // Adds the resolved layouts to all tensor types of `streamOps`, if encodings
+  // are present.
+  LogicalResult run();
+
+private:
+  // Appends the query from the `affinityOp` to `queries`. Note that most of
+  // operations only care the execution affinity. There are outliers (e.g.,
+  // tensor dispatch op, etc.) that need to resolve affinities for
+  // operand resources.
+  LogicalResult addQuery(IREE::Stream::AffinityAnalysis &affinityAnalysis,
+                         IREE::Stream::AffinityOpInterface affinityOp);
+
+  // The list of the queries that can be used for batch affinity queries. The
+  // analysis could be very expensive because it could apply the whole program
+  // data flow analysis.
+  SmallVector<IREE::Stream::AffinityAndOpPair> queries;
+
+  // The layout resolvers for each query.
+  llvm::DenseMap<IREE::Stream::AffinityAndOpPair, SetVector<Attribute>>
+      cachedLayoutAttrs;
+
+  // Input moduleOp. The op is not expected to be updated during the query.
+  // Because data flow analaysis can be involved. Modifying the IR invalidates
+  // the state and may lead to crashes as pointer references into the IR
+  // structure are retained.
+  ModuleOp moduleOp;
+
+  // The ops that need to be updated.
+  SmallVector<IREE::Stream::AffinityOpInterface> streamOps;
+
+  // The layout resolver function, which is used to resolve layouts for
+  // encodings. See StreamInterfaces.h for more details.
+  IREE::Stream::ResolveLayoutAttrFn resolveLayoutAttr;
+};
+} // namespace
+
+StreamTensorOpUpdater::StreamTensorOpUpdater(ModuleOp moduleOp)
```

Comment:
style: it's useful to keep functions inline unless it helps readability when the class is defined in the same file - it's fewer lines of code to have this up there and also harder to lose things (you declare `StreamTensorOpUpdater();` as well but never define it, which is harder to see because it's out-of-line)